### ISSUE TITLE
Transform fails with `Unexpected "..."`

### ISSUE THREAD

**Issue by IMax153** [vitest-dev/vitest#2157]
### Describe the bug

Upstream issue in `vite` - https://github.com/vitejs/vite/issues/10501

Annotating the following re-exported function with it's type:

```ts
/**
 * Constructs a tuple of results from a tuple of `SafeEval`s.
 *
 * @since 1.0.0
 * @category constructors
 */
export const tuple: <EN extends ReadonlyArray<SafeEval<any>>>(
  ...[e1, e2, ...es]: EN & {
    readonly 0: SafeEval<any>
    readonly 1: SafeEval<any>
  }
) => SafeEval<Readonly<{ [K in keyof EN]: [EN[K]] extends [SafeEval<infer A>] ? A : never }>> =
  SE.tuple
```

causes `vitest` to fail with the following error:

```
❯ failureErrorWithLog node_modules/.pnpm/esbuild@0.15.11/node_modules/esbuild/lib/main.js:1566:15
Error: Transform failed with 1 error:
/workspace/data/src/SafeEval.ts:86:2: ERROR: Unexpected "..."
❯ node_modules/.pnpm/esbuild@0.15.11/node_modules/esbuild/lib/main.js:805:29
❯ responseCallbacks.<computed> node_modules/.pnpm/esbuild@0.15.11/node_modules/esbuild/lib/main.js:671:9
❯ handleIncomingPacket node_modules/.pnpm/esbuild@0.15.11/node_modules/esbuild/lib/main.js:726:9
❯ Socket.readFromStdout node_modules/.pnpm/esbuild@0.15.11/node_modules/esbuild/lib/main.js:647:7
❯ Socket.emit node:events:513:28
```

The error references an `Unexpected "..."` in the type annotation (preceding the tuple parameter) even though this is perfectly valid typescript.

Removing the type annotation causes tests to pass successfully:

```ts
/**
 * Constructs a tuple of results from a tuple of `SafeEval`s.
 *
 * @since 1.0.0
 * @category constructors
 */
export const tuple = SE.tuple
```

### Reproduction

Navigate to the following GitPod url:

```
https://gitpod.io/#https://github.com/fp-ts/data
```

Add the type annotation to `tuple` in the `SafeEval` module [here](https://github.com/fp-ts/data/blob/f193b6a8deb0bdb97e25c29ef69f8ad67a1faca0/src/SafeEval.ts#L98).

Run `pnpm install` followed by `pnpm test`.

### System Info

```shell
System:
    OS: Linux 5.15 Ubuntu 20.04.5 LTS (Focal Fossa)
    CPU: (16) x64 AMD EPYC 7B13
    Memory: 41.06 GB / 62.80 GB
    Container: Yes
    Shell: 5.0.17 - /bin/bash
  Binaries:
    Node: 16.17.1 - ~/.nvm/versions/node/v16.17.1/bin/node
    Yarn: 1.22.19 - ~/.nvm/versions/node/v16.17.1/bin/yarn
    npm: 8.15.0 - ~/.nvm/versions/node/v16.17.1/bin/npm
  npmPackages:
    vite: ^2.9.13 => 2.9.15 
    vitest: 0.24.3 => 0.24.3
```

### Used Package Manager

pnpm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by sheremet-va** [comment]
Vitest doesn't transform files on it's own. It uses Vite. Please, open issue in Vite repo with the reproduction.

**Comment by IMax153** [comment]
@sheremet-va - added the upstream issue to the description

**Comment by sapphi-red** [comment]
This is fixed by esbuild [v0.17.9](https://github.com/evanw/esbuild/releases/tag/v0.17.9) and Vite v4.2.0.
